src/query/functions/temporal/rate.go

```diff
@@ -36,36 +36,167 @@ const (
 	// IDeltaType calculates the difference between the last two values in the time series.
 	// IDeltaTemporalType should only be used with gauges.
 	IDeltaType = "idelta"
-)
 
-// NewRateOp creates a new base temporal transform for rate functions
-func NewRateOp(args []interface{}, optype string) (transform.Params, error) {
-	if optype == IRateType || optype == IDeltaType {
-		return newBaseOp(args, optype, newRateNode, nil)
-	}
+	// RateType calculates the per-second average rate of increase of the time series.
+	RateType = "rate"
 
-	return nil, fmt.Errorf("unknown rate type: %s", optype)
+	// DeltaType calculates the difference between the first and last value of each time series.
+	DeltaType = "delta"
+
+	// IncreaseType calculates the increase in the time series.
+	IncreaseType = "increase"
+)
+
+type rateProcessor struct {
+	isRate, isCounter bool
+	rateFn            rateFn
 }
 
-func newRateNode(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
-	isRate := op.operatorType == IRateType
+func (r rateProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
+	var (
+		isRate, isCounter bool
+		rateFn            = standardRateFunc
+	)
+
+	switch op.operatorType {
+	case IRateType:
+		isRate = true
+		rateFn = irateFunc
+	case IDeltaType:
+		rateFn = irateFunc
+	case RateType:
+		isRate = true
+		isCounter = true
+	case IncreaseType:
+		isCounter = true
+	}
 
 	return &rateNode{
 		op:         op,
 		controller: controller,
 		timeSpec:   opts.TimeSpec,
 		isRate:     isRate,
+		isCounter:  isCounter,
+		rateFn:     rateFn,
 	}
 }
 
+// NewRateOp creates a new base temporal transform for rate functions
+func NewRateOp(args []interface{}, optype string) (transform.Params, error) {
+	if optype == IRateType || optype == IDeltaType || optype == RateType ||
+		optype == IncreaseType || optype == DeltaType {
+		r := rateProcessor{}
+		return newBaseOp(args, optype, r)
+	}
+
+	return nil, fmt.Errorf("unknown rate type: %s", optype)
+}
+
+type rateFn func([]float64, bool, bool, transform.TimeSpec) float64
+
 type rateNode struct {
-	op         baseOp
-	controller *transform.Controller
-	timeSpec   transform.TimeSpec
-	isRate     bool
+	op                baseOp
+	controller        *transform.Controller
+	timeSpec          transform.TimeSpec
+	isRate, isCounter bool
+	rateFn            rateFn
 }
 
 func (r *rateNode) Process(values []float64) float64 {
+	return r.rateFn(values, r.isRate, r.isCounter, r.timeSpec)
+}
+
+func standardRateFunc(values []float64, isRate, isCounter bool, timeSpec transform.TimeSpec) float64 {
+	var (
+		rangeStart = float64(timeSpec.Start.Unix()) - timeSpec.Step.Seconds()
```

Comment:
instead of converting all these into floats, lets keep them as time.Time or time.Duration as much as possible